rpmlb/recipe.py

```diff
@@ -23,22 +25,39 @@ def __init__(self, file_path, recipe_id):
         self.num_of_package = len(self.recipe['packages'])
 
     def each_normalized_package(self):
+        """Present recipe packages in normalized form.
+
+        Yields:
+            Dictionary with the package's data.
+
+        Common keys in the yielded dictionary:
+            - name: The package's name
+            - booststrap: Numerical position in the bootstrapping
+                sequence, or None if no bootstrap is necessary.
+            - macros, replaced_macros: Mapping of macro name to (new)
+                macro body.
+        """
+
         packages = self.recipe['packages']
+        bootstrap_map = self._count_bootstrap_sequences()
+
         for package in packages:
-            package_dict = {}
-            # String or hash
-            if isinstance(package, str):
-                package_dict['name'] = package
-            elif isinstance(package, dict):
-                keys = list(package.keys())
-                name = keys[0]
-                if not package[name]:
-                    raise ValueError('Recipe package %s is invalid format.' %
-                                     name)
-                package_dict = package[name]
-                package_dict['name'] = name
-            else:
-                raise ValueError('package is invalid.')
+            name = self._package_name(package)
+            bootstrap_position = next(bootstrap_map[name], None)
+
+            package_dict = {
+                'name': name,
+                'bootstrap': bootstrap_position,
+            }
+
+            if isinstance(package, Mapping):
```

Comment:
`isinstance(package, dict)` forces the package to be an actual built-in dictionary or subclass.
`isinstance(package, Mapping)` says "anything that acts/says it acts as a dictionary", which is more in line with the python duck-typing way.